ARCs/arc-0004.md

```diff
@@ -0,0 +1,498 @@
+   value at `a[i][k][l][r]`.
+2. The encoding of a value or array element is not interleaved with
+   other data and it is relocatable, i.e. only relative “addresses”
+   (indexes to other parts of the encoding) are used.
+
+
+### Types
+
+The following types are supported in the Algorand ABI.
+
+* `uint<N>`: An `N`-bit unsigned integer, where `8 <= N <= 512` and `N % 8 = 0`.
+* `byte`: An alias for `uint8`.
+* `bool`: A boolean value that is restricted to either 0 or 1. When encoded, up to 8 consecutive `bool` values will be packed into a single byte.
+* `ufixed<N>x<M>`: An `N`-bit unsigned fixed-point decimal number with precision `M`, where `8 <= N <= 512`, `N % 8 = 0`, and `0 < M <= 160`, which denotes a value `v` as `v / (10^M)`.
+* `<type>[<N>]`: A fixed-length array of length `N`, where `N >= 0`. `type` can be any other type.
+* `address`: Used to represent a 32-byte Algorand address. This is equivalent to `byte[32]`.
+* `<type>[]`: A variable-length array. `type` can be any other type.
+* `string`: A variable-length byte array (`byte[]`) assumed to contain UTF-8 encoded content.
+* `(T1,T2,...,TN)`: A tuple of the types `T1`, `T2`, …, `TN`, `N >= 0`.
+
+
+### Static vs Dynamic Types
+
+For encoding purposes, the types are divided into two categories: static and dynamic.
+
+The dynamic types are:
+
+*  `<type>[]` for any `type`
+    * This includes `string` since it is an alias for `byte[]`.
+* `<type>[<N>]` for any dynamic `type`
+* `(T1,T2,...,TN)` if `Ti` is dynamic for some `1 <= i <= N`
+
+All other types are static. For a static type, all encoded values of
+that type have the same length, irrepective of their actual value.
+
+
+### Encoding Rules
+
+Let `len(a)` be the number of bytes in the binary string `a`. The
+returned value shall be considered to have the ABI type `uint16`.
+
+Let `enc` be a mapping from values of the ABI types to binary
+strings. This mapping defines the encoding of the ABI.
+
+For any ABI value `x`, we recursively define `enc(x)` to be as follows:
+
+* If `x` is a tuple of `N` types, `(T1,T2,...,TN)`, where `x[i]` is the value at index `i`, starting at 1:
+    * `enc(x) = head(x[1]) ... head(x[N]) tail(x[1]) ... tail(x[N])`
+    * Let `head` and `tail` be mappings from values in this tuple to binary strings. For each `i` such that `1 <= i <= N`, these mappings are defined as:
+        * If `Ti` (the type of `x[i]`) is static:
+            * If `Ti` is `bool`:
+                * Let `after` be the largest integer such that all `T(i+j)` are `bool`, for `0 <= j <= after`.
+                * Let `before` be the largest integer such that all `T(i-j)` are `bool`, for `0 <= j <= before`.
+                * If `before % 8 == 0`:
+                    * `head(x[i]) = enc(x[i]) | (enc(x[i+1]) >> 1) | ... | (enc(x[i + min(after,7)]) >> min(after,7))`, where `>>` is bitwise right shift which pads with 0, `|` is bitwise or, and `min(x,y)` returns the minimum value of the integers `x` and `y`.
```

Comment:
The decision to encode boolean values "backwards" was made so that their encoding is consistent when either a single or multiple boolean values are grouped together. E.g. a single true is encoded as `0b10000000`, and two consecutive trues are encoded as `0b11000000`, rather than have the first boolean value relocate from the rightmost to leftmost bit (`0b00000001` -> `0b11000000`).

This decision was also influenced by the presence of the `getbit` and `setbit` opcodes which make it relatively easy for TEAL programs to manipulate individual bits, regardless of where they are in the byte array.

In your opinion what would be the more sensible option?

Reply:
With your current encoding scheme, I cannot write a compositional runtime or compile-time parser that is type directed, because I need to look at the adjacent values. (That is you cannot write a `encode(Array<elem, len>)(value)` that only has access to `encode(elem)` (as a function or a code generator) and not `elem` itself. In C++ terminology, the `encode` template has a special case for `boolean` that has to look at adjacent values in the sequence, which would be very complex to do.)

I think the natural thing is for `bytes encode([true, false, false]) ; getbit 0` to return `true`, because it is the first bool in the array. Of course, any compiler can just be programmed to "deal with it", but given that you have to make a comment in bold in this documentation... you *know* it is confusing. So stop yourself :)

My preference is simply to make the only special case `encode(Array<bool, N>)` (including dynamic `N`) and then `Tuple<bool, bool>` is different than `Array<bool, 2>` but that's fine and compositional. (You could, if you wanted, say something like "All sequences of `bool` types in tuples are collapsed to a single array" or even "It is illegal for `bool` types to appear adjacent in tuples" so people have to convert to array by hand.)

Reply:
>  In C++ terminology, the `encode` template has a special case for `boolean` that has to look at adjacent values in the sequence, which would be very complex to do.

Absolutely. We decided the additional encoding complexity here is worth the space savings.

One trick that can simplify encoding bools would be to have a preprocessor that replaces all adjacent bool values with a `BoolSequence` pseudo-type before encoding. This type would internally contain multiple boolean values (either up to 8, or an unlimited amount based on how you implement it) and it would be encodable without looking at neighboring types anymore.

> I think the natural thing is for `bytes encode([true, false, false]) ; getbit 0` to return `true`, because it is the first bool in the array.

I agree, and if you invoke `getbit 0` with the byte string `0b10000000`, you will in fact get 1.

> My preference is simply to make the only special case `encode(Array<bool, N>)` (including dynamic `N`) and then `Tuple<bool, bool>` is different than `Array<bool, 2>` but that's fine and compositional.

If we decide the current encoding system _is_ too complicated, this might be a good compromise.

Reply:
You are correct about `getbit` working with this encoding... but that's just because `getbit` is the source of the craziness. :) Ideally that would be fixed so that `int 0; itob; int 0; int 1; setbit; btoi; assert` is true. This encoding just doubles down on that mistake. (I think it is a mistake because setbit/getbit behave completely differently on arrays and numbers. We should think of bytes as just arrays of bits, so if I have `array(byte(a, b, c, d, e, f, g, h), byte(A, B, C, D, E, F, G, H))`, then `getbit 0` should be `a` and `getbit 7` `h` and `getbit 15` `H`, because if I had an array of arrays in C-style row-major order they'd be stored as `[a, b, c, d, e, f, g, h, A, B, C, D, E, F, G, H]` and `getbit`/`setbit` just do indexing into that array. However, TEAL is set up so that the array is `[ h, g, f, e, d, c, b, a, H, G, F, E, D, C, B, A ]`, which I believe is an attempt to match the way we typically type out bytes rather than what the bytes actually are---the zeroth bit is the one we type on the right, not the one we type on the left.

Reply:
Just to be super clear:
1. The comments in the implementation of setbit/etc say "This doesn't mean what you think it should mean and is backward".
2. The documentation for setbit/etc says the same thing: this is weird, let me show you an example so you see how weird it is and how different it is than when it operates on a uint64 (which you might think is just a byte[8])
3. This specification says how weird and unintuitive this encoding is

Put all those together and ask where the problem is: In the brains of the readers who expect the "wrong" thing? Or in the implementation?

Reply:
As a reader, I found myself reading this definition in reverse order, as it is easier to grok the base cases and then build up to the recursive case. So if I'm not an outlier, it might be worth putting the base cases before the recursive.